### DISCUSSION TITLE
Best practice for running/launching a server before tests?

### DISCUSSION THREAD

**Discussion by holtwick** [vitest-dev/vitest#334] [Q&A] [answered]
What would be the recommended solution to start up a server on which tests rely? 

Would it be using `npm-run-all` together with `vitest` from package.json?

Or is there some way to globally execute some JS/TS before running/watching tests, like in Playwright/Jest? 

https://playwright.dev/docs/test-advanced#global-setup-and-teardown
https://jestjs.io/docs/configuration#globalsetup-string

**Comment by patak-cat** [comment, marked as answer]
There isn't yet an API for this in Vitest, see https://github.com/vitest-dev/vitest/issues/300#issuecomment-1001472631, and discussion in Discord started by @dominikg here https://discord.com/channels/917386801235247114/924960069039325224/924960918373290014
@dominikg suggested that a possible solution would be to use Vite plugins with the `buildStart` and `buildEnd` hooks, that could be an interesting way to avoid complexity in Vitest

**Reply by holtwick** [reply]
Thanks @patak-dev for pointing me to the discussion. `beforeAll` and `setupFiles` are executed before each test as far as I know. I'm referring to a situation, where testing depends on a running service, which does not always has to be `vite` only. Maybe an example says more then thousand words ;) Here is a similar script I use with Playwright:

```ts
import { FullConfig } from "@playwright/test"
import { resolve } from "path"
import { createServer } from "vite"

async function globalSetup(config: FullConfig) {
  // https://vitejs.dev/guide/api-javascript.html#createserver
  const server = await createServer({
    root: resolve(__dirname, ".."),
    server: {
      port: 8080,
    },
  })

  await server.listen()

  return async () => {
    await server.close()
  }
}

export default globalSetup
```

The main advantages are:

- Integrated with the lifetime cycle of the testing environment
- Is not restarted on each test or during vitest's watch period
- Gracefully tears down

Anyway, I got it working with [concurrently](https://www.npmjs.com/package/concurrently) for my use case for now, like this (package.json):

```json
 "fulltest": "concurrently -k -r -i \"pnpm start\" \"pnpm run test\"",
```

Thanks for your help and last but not least: VITEST IS AWESOME! :)

**Reply by patak-cat** [reply]
Yes, these discussion that I linked are about global teardown and setup. We need this to port the vite test suite to vitest, we are also using playwright there

**Reply by holtwick** [reply]
Thanks, that sounds good.

**Reply by tirli** [reply]
@patak-dev Do you have any ideas on how to expose the actual port to the tests? Since vite is having a feature like "Port 3000 is in use, trying another one..." and it could end up using the wrong port.
I'm also using vite + playwright

**Comment by dominikg** [comment]
this change should allow buildEnd hook to work in vitest watch mode too https://github.com/vitest-dev/vitest/pull/349

after that next step would be to create a `vite-plugin-playwright-server` that creates the browser server on buildStart and exposes it via an api. Please join vitest discord if you want to collaborate on that. (or find me on vite discord)

**Reply by holtwick** [reply]
Thanks for the invitation to discord. I'm sure all paths described are leading into the right direction. Thanks a lot!

**Comment by hubertgrzeskowiak** [comment]
Found this thread while trying to test a client-server app. I ended up launching the server in a separate process in the background before starting vitest in browser mode:


package.json
```
{
  "scripts": {
    "test": "rm -rf echo.log; vite-node echo-server.ts &> echo.log & vitest --run; kill %1; echo 'Background echo-server logs:'; cat echo.log"
    ...
```

I wish, there was a way to start vitest programmatically...

**Comment by hubertgrzeskowiak** [comment]
After reading https://github.com/vitest-dev/vitest/issues/524 I just found this nice solution!

```
import { createVitest } from "vitest/node";

// server init code goes here
// ...

const ctx = await createVitest("test", {
  watch: false
});
await ctx.start();
await ctx.close();
```

Then you can run this script with `vite-node test.ts`
